File: plugins/reverb/gated-reverb/GatedReverb.java
package com.vocalmonitor.plugin.community;

import com.vocalmonitor.plugin.VocalMonitorNativePlugin;

// Gated Reverb (native port) — 6-comb plate reverb followed by an
// envelope-triggered hard gate with hold. The Phil Collins / Hugh
// Padgham 80s drum sound on a vocal.
public final class GatedReverb implements VocalMonitorNativePlugin {
    private static final int[] COMB_BASE = { 1687, 1601, 2053, 2251, 1499, 1789 };
    private static final int[] AP_BASE = { 347, 113, 41 };
    private float[][] combs = new float[6][];
    private final int[] combLen = new int[6];
    private final int[] combIdx = new int[6];
    private final float[] combLp = new float[6];
    private float[][] aps = new float[3][];
    private final int[] apLen = new int[3];
    private final int[] apIdx = new int[3];
    private float env = 0f, gateGain = 0f;
    private int holdSamples = 0;
    private int sampleRate = 44100;
    private float decay = 0.85f, threshold = 0.03f, hold = 220f, release = 30f, mix = 0.5f;

    @Override
    public void init(int sr) {
        this.sampleRate = sr;
        float srScale = sr / 44100f;
        for (int c = 0; c < 6; c++) {
            int L = (int) Math.ceil(COMB_BASE[c] * srScale);
            combs[c] = new float[L];
            combLen[c] = L;
            combIdx[c] = 0;
            combLp[c] = 0f;
        }
        for (int a = 0; a < 3; a++) {
            int L = (int) Math.ceil(AP_BASE[a] * srScale);
            aps[a] = new float[L];
            apLen[a] = L;
            apIdx[a] = 0;
        }
        env = gateGain = 0f;
        holdSamples = 0;
    }

    @Override public String[] parameterNames() {
        return new String[] { "decay", "threshold", "hold", "release", "mix" };
    }
    @Override public float parameterMin(String n) {
        switch (n) {
            case "decay":     return 0.5f;
            case "threshold": return 0.001f;
            case "hold":      return 30f;
            case "release":   return 1f;
            default:          return 0f;
        }
    }
    @Override public float parameterMax(String n) {
        switch (n) {
            case "decay":     return 0.97f;
            case "threshold": return 0.3f;
            case "hold":      return 800f;
            case "release":   return 300f;
            default:          return 1f;
        }
    }
    @Override public float parameterDefault(String n) {
        switch (n) {
            case "decay":     return 0.85f;
            case "threshold": return 0.03f;
            case "hold":      return 220f;
            case "release":   return 30f;
            case "mix":       return 0.5f;
            default:          return 0f;
        }
    }
    @Override public String parameterLabel(String n) {
        switch (n) {
            case "decay":     return "Decay";
            case "threshold": return "Threshold";
            case "hold":      return "Hold (ms)";
            case "release":   return "Rel (ms)";
            case "mix":       return "Mix";
            default:          return n;
        }
    }
    @Override public void setParameter(String n, float v) {
        switch (n) {
            case "decay":     decay = v; break;
            case "threshold": threshold = v; break;
            case "hold":      hold = v; break;
            case "release":   release = v; break;
            case "mix":       mix = v; break;
        }
    }

    @Override
    public void process(float[] input, float[] output) {
        final float envAttack = 1f - (float) Math.exp(-1.0 / (sampleRate * 0.005));
        final float envRelease = 1f - (float) Math.exp(-1.0 / (sampleRate * 0.05));
        final int holdN = (int) Math.floor(hold * sampleRate / 1000f);
        final float relCoef = 1f - (float) Math.exp(-1.0 / Math.max(1, sampleRate * release / 1000.0));
        final float apG = 0.5f;
        final float decayLocal = decay;
        final float threshLocal = threshold;
        final float mixLocal = mix;
        final float dry = 1f - mixLocal;
        final int[] cls = new int[6];
        for (int c = 0; c < 6; c++) cls[c] = combLen[c] - 2;
        float e = env, gg = gateGain;
        int hs = holdSamples;
        final int n = input.length;
        for (int i = 0; i < n; i++) {
            float x = input[i];
            float rect = x < 0 ? -x : x;
            float coef = rect > e ? envAttack : envRelease;
            e = e + coef * (rect - e);
            float target;
            if (e > threshLocal) { target = 1f; hs = holdN; }
            else if (hs > 0) { target = 1f; hs--; }
            else target = 0f;
            gg = gg + relCoef * (target - gg);

            float rvIn = x;
            float sum = 0f;
            for (int c = 0; c < 6; c++) {
                float[] b = combs[c];
                int idx = combIdx[c];
                int r = idx - cls[c]; if (r < 0) r += combLen[c];
                float d = b[r];
                combLp[c] = combLp[c] + 0.35f * (d - combLp[c]);
                b[idx] = rvIn + combLp[c] * decayLocal;
                idx++; if (idx >= combLen[c]) idx = 0;
                combIdx[c] = idx;
                sum += combLp[c];
            }
            float v = sum * (1f / 6f);
            for (int a = 0; a < 3; a++) {
                float[] ab = aps[a];
                int aIdx = apIdx[a];
                int aL = apLen[a] - 2;
                int ar = aIdx - aL; if (ar < 0) ar += apLen[a];
                float ad = ab[ar];
                float ai = v + ad * apG;
                ab[aIdx] = ai;
                v = ad - ai * apG;
                aIdx++; if (aIdx >= apLen[a]) aIdx = 0;
                apIdx[a] = aIdx;
            }
            float wet = v * gg;
            output[i] = x * dry + wet * mixLocal;
        }
        env = e; gateGain = gg; holdSamples = hs;
    }
}

File: plugins/reverb/gated-reverb/gated-reverb.js
// Gated Reverb — the trademark 80s drum sound (Phil Collins "In The Air
// Tonight", every snare on Peter Gabriel III). A dense plate-style
// reverb is followed by an envelope-triggered noise gate: while the
// input is loud, the reverb tail is wide open; when the input drops
// below a threshold, the wet path is hard-cut after a short hold time.
// The result is huge, dense, then suddenly silent — no natural decay.

function GatedReverbProcessor() {
  var srScale = sampleRate / 44100;
  this.srScale = srScale;
  // Schroeder core. Bigger combs than plate-reverb for that ambient bigness.
  this.combBase = [1687, 1601, 2053, 2251, 1499, 1789];
  this.combs = [];
  this.combLen = [];
  this.combIdx = [0, 0, 0, 0, 0, 0];
  this.combLp = [0, 0, 0, 0, 0, 0];
  for (var c = 0; c < 6; c++) {
    var L = Math.ceil(this.combBase[c] * srScale);
    var b = new Array(L);
    for (var k = 0; k < L; k++) b[k] = 0;
    this.combs.push(b);
    this.combLen.push(L);
  }
  this.apBase = [347, 113, 41];
  this.aps = []; this.apLen = []; this.apIdx = [0, 0, 0];
  for (var a = 0; a < 3; a++) {
    var L = Math.ceil(this.apBase[a] * srScale);
    var b = new Array(L);
    for (var k = 0; k < L; k++) b[k] = 0;
    this.aps.push(b);
    this.apLen.push(L);
  }
  // Envelope-follower state, gate state machine ("open"/"hold"/"closing").
  this.env = 0;
  this.gateGain = 0;
  this.holdSamples = 0;
  // Read-pointer offset into a small ringbuffer so the gate looks at the
  // INPUT not the reverb tail (otherwise the tail re-triggers itself).
}

GatedReverbProcessor.parameterDescriptors = [
  { name: 'decay',     label: 'Decay',     defaultValue: 0.85, minValue: 0.5, maxValue: 0.97 },
  { name: 'threshold', label: 'Threshold', defaultValue: 0.03, minValue: 0.001, maxValue: 0.3 },
  { name: 'hold',      label: 'Hold (ms)', defaultValue: 220,  minValue: 30,  maxValue: 800 },
  { name: 'release',   label: 'Rel (ms)',  defaultValue: 30,   minValue: 1,   maxValue: 300 },
  { name: 'mix',       label: 'Mix',       defaultValue: 0.5,  minValue: 0,   maxValue: 1 }
];

GatedReverbProcessor.prototype.process = function (inputs, outputs, parameters) {
  var input = inputs[0][0], output = outputs[0][0];
  if (!input || !output) return true;
  var decay = parameters.decay[0];
  var threshold = parameters.threshold[0];
  var holdMs = parameters.hold[0];
  var relMs = parameters.release[0];
  var mix = parameters.mix[0];

  var envAttack = 1 - Math.exp(-1 / (sampleRate * 0.005));
  var envRelease = 1 - Math.exp(-1 / (sampleRate * 0.05));
  var holdN = Math.floor(holdMs * sampleRate / 1000);
  var relCoef = 1 - Math.exp(-1 / Math.max(1, sampleRate * relMs / 1000));
  var apG = 0.5;
  // Comb size, full buffer.
  var cls = [this.combLen[0]-2, this.combLen[1]-2, this.combLen[2]-2, this.combLen[3]-2, this.combLen[4]-2, this.combLen[5]-2];

  for (var i = 0; i < input.length; i++) {
    var x = input[i];

    // --- Envelope follower on the DRY input. ---
    var rect = x < 0 ? -x : x;
    var coef = rect > this.env ? envAttack : envRelease;
    this.env = this.env + coef * (rect - this.env);

    // Gate state machine: open while above threshold, then hold, then
    // close exponentially. This is what makes it sound 80s.
    var target;
    if (this.env > threshold) {
      target = 1;
      this.holdSamples = holdN;
    } else if (this.holdSamples > 0) {
      target = 1;
      this.holdSamples--;
    } else {
      target = 0;
    }
    this.gateGain = this.gateGain + relCoef * (target - this.gateGain);

    // --- Reverb core (6 combs + 3 allpasses). ---
    var rvIn = x;
    var sum = 0;
    for (var c = 0; c < 6; c++) {
      var b = this.combs[c];
      var idx = this.combIdx[c];
      var r = idx - cls[c]; if (r < 0) r += this.combLen[c];
      var d = b[r];
      this.combLp[c] = this.combLp[c] + 0.35 * (d - this.combLp[c]);
      b[idx] = rvIn + this.combLp[c] * decay;
      idx++; if (idx >= this.combLen[c]) idx = 0;
      this.combIdx[c] = idx;
      sum += this.combLp[c];
    }
    var v = sum * (1 / 6);
    for (var a = 0; a < 3; a++) {
      var ab = this.aps[a];
      var aIdx = this.apIdx[a];
      var aL = this.apLen[a] - 2;
      var ar = aIdx - aL; if (ar < 0) ar += this.apLen[a];
      var ad = ab[ar];
      var ai = v + ad * apG;
      ab[aIdx] = ai;
      v = ad - ai * apG;
      aIdx++; if (aIdx >= this.apLen[a]) aIdx = 0;
      this.apIdx[a] = aIdx;
    }

    var wet = v * this.gateGain;
    output[i] = x * (1 - mix) + wet * mix;
  }
  return true;
};

registerProcessor('gated-reverb', GatedReverbProcessor);
